sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -549,6 +549,20 @@ gather(AccessorTy acc,
 #endif
 }
 
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+template <typename T, int N, typename AccessorTy, typename Toffset>
+__ESIMD_API std::enable_if_t<
+    (sizeof(T) <= 4) && (N == 1 || N == 8 || N == 16 || N == 32) &&
+        !std::is_pointer<AccessorTy>::value && std::is_integral_v<Toffset> &&
+        !std::is_same_v<Toffset, uint64_t>,
+    simd<T, N>>
+gather(AccessorTy acc, simd<Toffset, N> offsets, uint64_t glob_offset = 0,
```

Comment:
the reason we need these functions is that if the user has a simd object that is not uint64_t, conversion may fail depending on the explicitly specified template arguments. see an example here:
https://godbolt.org/z/fcPz1aEPf

this was caught from a test where we were passing uint32_t and supplied no template args (so just like gather(acc ...). Most of our tests provide template args so it happens to work.

if the compiler needs to match, it won't call the conversion constructor.

but we need to be careful with this function:
1) we need to not call it if the type is uint64_t, otherwise we get a ambiguous functions
2) we need to not call it for scalar, otherwise we get abiguous calls. since we have Toffset, we won't match with this version for scalar and will call the other one.